dynamiqs/sparse.py

```diff
@@ -0,0 +1,149 @@
+    def __init__(self, diags: Array, offsets: tuple[int]):
+        self.offsets = offsets
+        self.diags = diags
+
+
+    def to_dense(self) -> Array:
+        """Turn any set of diagonals & offsets in a full NxN matrix."""
+        N = self.diags.shape[1]
+        out = jnp.zeros((N, N))
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, offset)
+            end = min(N, N + offset)
+            out += jnp.diag(diag[start:end], k=offset)
+        return out
+    
+    def to_sparse(self, matrix) -> Array:
+        """Turn a NxN sparse matrix into the amazing SparseDIA format"""
+        diagonals_with_offsets = {}
+        
+        # there are 2N-1 offsets in a NxN matrix
+        offset_range = 2 * matrix.shape[0] - 1
+        offset_center = offset_range // 2
+
+        for offset in range(-offset_center, offset_center + 1):
+            diagonal = jnp.diagonal(matrix, offset=offset)
+            if jnp.any(diagonal != 0):
+                diagonals_with_offsets[offset] = diagonal
+
+        return diagonals_with_offsets
+
+
+    @functools.partial(jax.jit, static_argnums=(0,1))
+    def _matmul(
+        self,
+        direction: str,
+        matrix: Array,  # If matrix is a vector need to input as column vector.
+    ) -> Array:
+        """Sparse @ Dense only using information about diagonals & offsets."""
+        N = matrix.shape[0]
+        out = jnp.zeros_like(matrix)
+        for offset, diag in zip(self.offsets, self.diags):
+            start = max(0, offset)
+            end = min(N, N + offset)
+            top = max(0, -offset)
+            bottom = top + end - start
+            if direction=="left":
+                out = out.at[top:bottom, :].add(
+                    diag[start:end, None] * matrix[start:end, :]
+                )
+            else:
+                out = out.at[:, start:end].add(
+                    jnp.transpose(diag[start:end, None] * jnp.transpose(matrix[:, top:bottom]))
+                )
+        return out
+    
+    @functools.partial(jax.jit, static_argnums=(0,1))
+    def _diamul(self, matrix) -> Array:
+        N = matrix.diags.shape[1]
+        
+        vector_list = []
+        offset_list = []
+
+        for offset, diag in zip(self.offsets, self.diags):
+            for matrix_offset, matrix_diag in zip(matrix.offsets, matrix.diags):
+
+                out_diag = jnp.zeros_like(diag)
+                
+                sA, sB = max(0, -matrix_offset), max(0, matrix_offset)
+                eA, eB = min(N, N-matrix_offset), min(N, N+matrix_offset)
+
+                out_diag = out_diag.at[sB:eB].add(
+                    diag[sA:eA] * matrix_diag[sB:eB]
+                )
+
+                new_offset = offset + matrix_offset
+
+                vector_list.append(out_diag)
+                offset_list.append(new_offset)
+        
+        out = jnp.vstack(vector_list)
+        
+        return out, offset_list 
+
+        
+    
+
+    def _dag(self):
```

Comment:
I think this should rather be `def dag(self) -> SparseDIA` and return an instance of the hermitian instance of the class.